system/include/emscripten/dom_pk_codes.h

```diff
@@ -0,0 +1,506 @@
+#define DOM_PK_EJECT                0xE02C /* "Eject"              */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "AudioVolumeDown"    */
+#define DOM_PK_AUDIO_VOLUME_DOWN    0xE02E /* "VolumeDown"         */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "AudioVolumeUp"      */
+#define DOM_PK_AUDIO_VOLUME_UP      0xE030 /* "VolumeUp"           */
+#define DOM_PK_BROWSER_HOME         0xE032 /* "BrowserHome"        */
+#define DOM_PK_NUMPAD_DIVIDE        0xE035 /* "NumpadDivide"       */
+#define DOM_PK_ALT_RIGHT            0xE038 /* "AltRight"           */
+#define DOM_PK_HELP                 0xE03B /* "Help"               */
+#define DOM_PK_NUM_LOCK             0xE045 /* "NumLock"            */
+#define DOM_PK_HOME                 0xE047 /* "Home"               */
+#define DOM_PK_ARROW_UP             0xE048 /* "ArrowUp"            */
+#define DOM_PK_PAGE_UP              0xE049 /* "PageUp"             */
+#define DOM_PK_ARROW_LEFT           0xE04B /* "ArrowLeft"          */
+#define DOM_PK_ARROW_RIGHT          0xE04D /* "ArrowRight"         */
+#define DOM_PK_END                  0xE04F /* "End"                */
+#define DOM_PK_ARROW_DOWN           0xE050 /* "ArrowDown"          */
+#define DOM_PK_PAGE_DOWN            0xE051 /* "PageDown"           */
+#define DOM_PK_INSERT               0xE052 /* "Insert"             */
+#define DOM_PK_DELETE               0xE053 /* "Delete"             */
+#define DOM_PK_META_LEFT            0xE05B /* "MetaLeft"           */
+#define DOM_PK_OS_LEFT              0xE05B /* "OSLeft"             */
+#define DOM_PK_META_RIGHT           0xE05C /* "MetaRight"          */
+#define DOM_PK_OS_RIGHT             0xE05C /* "OSRight"            */
+#define DOM_PK_CONTEXT_MENU         0xE05D /* "ContextMenu"        */
+#define DOM_PK_POWER                0xE05E /* "Power"              */
+#define DOM_PK_BROWSER_SEARCH       0xE065 /* "BrowserSearch"      */
+#define DOM_PK_BROWSER_FAVORITES    0xE066 /* "BrowserFavorites"   */
+#define DOM_PK_BROWSER_REFRESH      0xE067 /* "BrowserRefresh"     */
+#define DOM_PK_BROWSER_STOP         0xE068 /* "BrowserStop"        */
+#define DOM_PK_BROWSER_FORWARD      0xE069 /* "BrowserForward"     */
+#define DOM_PK_BROWSER_BACK         0xE06A /* "BrowserBack"        */
+#define DOM_PK_LAUNCH_APP_1         0xE06B /* "LaunchApp1"         */
+#define DOM_PK_LAUNCH_MAIL          0xE06C /* "LaunchMail"         */
+#define DOM_PK_LAUNCH_MEDIA_PLAYER  0xE06D /* "LaunchMediaPlayer"  */
+#define DOM_PK_MEDIA_SELECT         0xE06D /* "MediaSelect"        */
+
+/* Maps the EmscriptenKeyboardEvent::code field from emscripten/html5.h to one of the DOM_PK codes above. */
+static DOM_PK_CODE_TYPE emscripten_compute_dom_pk_code(const char *keyCodeString)
+{
+  if (!keyCodeString) return 0;
+
+  /* Compute the collision free hash. */
+  unsigned int hash = 0;
+  while(*keyCodeString) hash = ((hash ^ 0x5EE148E8U) << 3) ^ (unsigned int)*keyCodeString++;
```

Comment:
> emscripten_compute_dom_pk_code not going to be called by users, but just internally?

No, the opposite. Internally we do not call this, or the `emscripten_dom_vk_to_string` function ever, but they are exposed for application developers to use. `emscripten_dom_vk_to_string` adds a considerable amount of data strings to the built application, so Emscripten runtime itself does not call that, but allows users to do so, e.g. in game keyboard mapping menus and similar, if they chooset to. Also `emscripten_compute_dom_pk_code` is "derived" information from the event structure, so we don't call that by default, but let users transform the `code` field to this enum if they want to use the enums. Nothing prevents applications from utilizing the `code` string directly in their applications, in which case neither of these functions are compiled into the code.